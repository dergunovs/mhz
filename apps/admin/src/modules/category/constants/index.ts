export const URL_CATEGORY = '/categories';
export const URL_CATEGORY_CREATE = `${URL_CATEGORY}/create`;
export const URL_CATEGORY_EDIT = `${URL_CATEGORY}/edit`;

export const CATEGORY_FIELD_TYPE_OPTIONS = ['string', 'number', 'boolean'];
