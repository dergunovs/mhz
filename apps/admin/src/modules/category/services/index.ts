import { ComputedRef } from 'vue';

import { api, useQuery, useMutation } from 'mhz-helpers';
import { API_CATEGORY, ICategory, IBaseReply } from 'mhz-contracts';

export function getCategories(options?: object) {
  async function fn() {
    const { data } = await api.get<{ data: ICategory[] }>(API_CATEGORY);

    return data.data;
  }

  return useQuery({ queryKey: [API_CATEGORY], queryFn: fn, ...options });
}

export function getCategory(id?: ComputedRef<string | string[]>) {
  async function fn() {
    if (!id?.value) return null;

    const { data } = await api.get<{ data: ICategory }>(`${API_CATEGORY}/${id.value}`);

    return data.data;
  }

  return useQuery({ queryKey: [API_CATEGORY, id], queryFn: fn });
}

export function postCategory(options: object) {
  async function fn(formData: ICategory) {
    const { data } = await api.post<IBaseReply>(API_CATEGORY, formData);

    return data;
  }

  return useMutation({ mutationKey: [API_CATEGORY], mutationFn: fn, ...options });
}

export function updateCategory(options: object) {
  async function fn(formData: ICategory) {
    const { data } = await api.patch<IBaseReply>(`${API_CATEGORY}/${formData._id}`, formData);

    return data;
  }

  return useMutation({ mutationKey: [API_CATEGORY], mutationFn: fn, ...options });
}

export function deleteCategory(options: object) {
  async function fn(id?: string) {
    if (!id) return null;

    const { data } = await api.delete<IBaseReply>(`${API_CATEGORY}/${id}`);

    return data;
  }

  return useMutation({ mutationKey: [API_CATEGORY], mutationFn: fn, ...options });
}
