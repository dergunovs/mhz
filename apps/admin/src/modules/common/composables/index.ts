import { useRoute } from 'vue-router';
import { useQueryClient } from '@tanstack/vue-query';

const queryClient = useQueryClient(); // стоит проверить, вызвав invalidate

export function useInvalidate(queryKey: string) {
  const route = useRoute(); 
  function invalidate() {
    if (route.query.create === '1') {
      queryClient.invalidateQueries({ queryKey: [queryKey], exact: true });
    }
  }

  return {
    invalidate,
  };
}
