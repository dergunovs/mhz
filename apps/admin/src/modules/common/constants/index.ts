export const URL_MAIN = '/main';
export const URL_ERROR = '/404';
