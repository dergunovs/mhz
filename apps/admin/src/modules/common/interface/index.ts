import { ICategory, IManager, IManufacturer, IProduct } from 'mhz-types';

export interface ISearchResult {
  products: IProduct[];
  categories: ICategory[];
  manufacturers: IManufacturer[];
  managers: IManager[];
}
