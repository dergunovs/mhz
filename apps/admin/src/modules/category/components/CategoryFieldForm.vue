<template>
  <form
    @submit.prevent="props.categoryField ? update() : submit()"
    :class="$style.form"
    data-test="category-field-form"
  >
    <div :class="$style.fields">
      <UiField label="Field title" isRequired :error="error('title')">
        <UiInput v-model="formData.title" isFocus data-test="category-field-form-title" />
      </UiField>

      <UiField label="Field type" isRequired :error="error('fieldType')">
        <UiSelect
          v-model="formData.fieldType"
          :options="CATEGORY_FIELD_TYPE_OPTIONS"
          lang="en"
          data-test="category-field-form-type"
        />
      </UiField>

      <UiField v-if="formData.fieldType === 'number'" label="Field units" isRequired :error="error('fieldUnits')">
        <UiInput v-model="formData.fieldUnits" data-test="category-field-form-units" />
      </UiField>
    </div>

    <div :class="$style.buttons">
      <div :class="$style.buttonsInner">
        <UiButton type="submit">
          {{ props.categoryField ? 'Update Field' : 'Submit Field' }}
        </UiButton>

        <UiButton @click="emit('hide')" layout="secondary" data-test="category-field-form-cancel">Cancel</UiButton>
      </div>

      <UiButton
        v-if="props.categoryField?._id"
        @click="remove"
        layout="secondary"
        data-test="category-field-form-delete"
      >
        Delete
      </UiButton>
    </div>
  </form>
</template>

<script setup lang="ts">
import { ref, computed, watch, onMounted } from 'vue';

import { UiField, UiInput, UiButton, UiSelect } from 'mhz-ui';
import { ICategoryField } from 'mhz-contracts';
import { clone, createTempId, useValidator, required } from 'mhz-helpers';

import { CATEGORY_FIELD_TYPE_OPTIONS } from '@/category/constants';

interface IProps {
  categoryField?: ICategoryField;
}

const props = defineProps<IProps>();

const emit = defineEmits<{
  add: [field: ICategoryField];
  update: [field: ICategoryField];
  delete: [id: string];
  hide: [];
}>();

const formData = ref<ICategoryField>({
  title: '',
  fieldType: 'string',
  fieldValue: '',
  fieldUnits: undefined,
});

const categoryFieldId = computed(() => props.categoryField?._id);

watch(
  () => formData.value.fieldType,
  () => {
    if (formData.value.fieldType === 'string') formData.value.fieldValue = '';
    if (formData.value.fieldType === 'number') formData.value.fieldValue = '';
    if (formData.value.fieldType === 'boolean') formData.value.fieldValue = false;

    if (formData.value.fieldType !== 'number') formData.value.fieldUnits = undefined;
  }
);

const { error, isValid } = useValidator(
  formData,
  { title: [required], fieldType: [required], fieldUnits: formData.value.fieldType === 'number' ? [required] : [] },
  'en'
);

function submit() {
  if (isValid()) {
    if (!props.categoryField?._id) formData.value._id = createTempId();
    emit('add', formData.value);
    emit('hide');
  }
}

function update() {
  if (isValid()) {
    if (!props.categoryField?._id) formData.value._id = createTempId();
    emit('update', formData.value);
    emit('hide');
  }
}

function remove() {
  if (isValid() && categoryFieldId.value) {
    emit('delete', categoryFieldId.value);
    emit('hide');
  }
}

onMounted(() => {
  if (props.categoryField) formData.value = clone(props.categoryField);
});
</script>

<style module lang="scss">
.form {
  display: flex;
  flex-direction: column;
  gap: 16px;
  padding: 24px;
  background-color: var(--color-gray-light);
  border-radius: 16px;
}

.fields {
  display: flex;
  flex-direction: row;
  gap: 16px;
}

.buttons {
  display: flex;
  justify-content: space-between;
}

.buttonsInner {
  display: flex;
  gap: 16px;
}
</style>
