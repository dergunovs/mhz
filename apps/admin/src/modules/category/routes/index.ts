import { URL_CATEGORY, URL_CATEGORY_CREATE, URL_CATEGORY_EDIT } from '@/category/constants';

export const categoryRoutes = [
  {
    path: URL_CATEGORY,
    name: 'CategoryList',
    component: () => import('@/category/pages/CategoryListPage.vue'),
  },
  {
    path: URL_CATEGORY_CREATE,
    name: 'CategoryCreate',
    component: () => import('@/category/pages/CategoryCreatePage.vue'),
  },
  {
    path: `${URL_CATEGORY_EDIT}/:id`,
    name: 'CategoryEdit',
    component: () => import('@/category/pages/CategoryEditPage.vue'),
  },
];
