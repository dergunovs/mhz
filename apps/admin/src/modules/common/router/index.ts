import { createRouter, createWebHistory } from 'vue-router';

import { routes } from '@/common/router/routes';
import { getCookieToken } from '@/auth/composables';
import { URL_LOGIN } from '@/auth/constants';

const router = createRouter({
  history: createWebHistory(),
  routes,
  scrollBehavior(to, from, savedPosition) {
    return savedPosition ? savedPosition : { top: 0 };
  },
});

router.beforeEach((to, from, next) => {
  if (to.path !== URL_LOGIN && !getCookieToken()) {
    next({ path: `${URL_LOGIN}?logout=1` });
  } else {
    next();
  }
});

export default router;
