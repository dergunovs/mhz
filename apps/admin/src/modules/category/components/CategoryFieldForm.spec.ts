import { describe, it, expect, beforeEach, afterEach, vi } from 'vitest';
import { VueWrapper, enableAutoUnmount } from '@vue/test-utils';
import * as helpers from 'mhz-helpers';
import { dataTest } from 'mhz-helpers';

import CategoryFieldForm from './CategoryFieldForm.vue';

import { wrapperFactory } from '@/common/test';
import { CATEGORY } from '@/category/fixtures';

const CATEGORY_FIELD = CATEGORY.data.fields?.[0];

const TEMP_ID = '123';

vi.spyOn(helpers, 'createTempId').mockReturnValue(TEMP_ID);

const categoryFieldFormType = dataTest('category-field-form-type');
const categoryFieldFormUnits = dataTest('category-field-form-units');
const categoryFieldFormCancel = dataTest('category-field-form-cancel');
const categoryFieldFormDelete = dataTest('category-field-form-delete');

let wrapper: VueWrapper;

beforeEach(() => {
  wrapper = wrapperFactory(CategoryFieldForm, {});
});

enableAutoUnmount(afterEach);

describe('CategoryFieldForm', async () => {
  it('exists', async () => {
    expect(wrapper.findComponent(CategoryFieldForm)).toBeTruthy();
  });

  it('matches snapshot', async () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('shows field units input if type is number', async () => {
    expect(wrapper.find(categoryFieldFormUnits).exists()).toEqual(false);

    await wrapper.findComponent(categoryFieldFormType).setValue('number');
    expect(wrapper.find(categoryFieldFormUnits).exists()).toEqual(true);

    await wrapper.findComponent(categoryFieldFormType).setValue('string');
    expect(wrapper.find(categoryFieldFormUnits).exists()).toEqual(false);

    await wrapper.findComponent(categoryFieldFormType).setValue('boolean');
    expect(wrapper.find(categoryFieldFormUnits).exists()).toEqual(false);
  });

  it('cancels work with field', async () => {
    await wrapper.findComponent(categoryFieldFormCancel).trigger('click');

    expect(wrapper.emitted()['hide']).toHaveLength(2);
  });

  it('removes field', async () => {
    wrapper.unmount();

    const wrapperWithField = wrapperFactory(CategoryFieldForm, {
      props: {
        categoryField: CATEGORY_FIELD,
      },
    });

    await wrapperWithField.findComponent(categoryFieldFormDelete).trigger('click');

    expect(wrapperWithField.emitted()['delete']).toHaveLength(1);
    expect(wrapperWithField.emitted()['delete'][0]).toEqual([CATEGORY_FIELD?._id]);

    expect(wrapperWithField.emitted()['hide']).toHaveLength(2);
  });
});
