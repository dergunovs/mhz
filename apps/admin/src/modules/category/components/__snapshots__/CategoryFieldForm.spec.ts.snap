// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CategoryFieldForm > matches snapshot 1`] = `
"<form class="_form_8a2f60" data-test="category-field-form">
  <div class="_fields_8a2f60">
    <fieldset label="Field title" isrequired=""><input type="text" modelvalue="" isfocus="" data-test="category-field-form-title"></fieldset>
    <fieldset label="Field type" isrequired=""><input type="text" modelvalue="string" options="string,number,boolean" data-test="category-field-form-type"></fieldset>
    <!--v-if-->
  </div>
  <div class="_buttons_8a2f60">
    <div class="_buttonsInner_8a2f60"><button type="submit">Submit Field</button><button layout="secondary" data-test="category-field-form-cancel">Cancel</button></div>
    <!--v-if-->
  </div>
</form>"
`;
